### DISCUSSION TITLE
How to send concurrent message to client on single connection SignalR

### DISCUSSION THREAD

**Discussion by quytrieu37** [dotnet/aspnetcore#54940] [Q&A]
Hello SignalR Team,
Our application use SignalR to retrieve data from a hub. Each client establishes a single connection at a time.
On the client side, we can invoke multiple methods on the server concurrently by adjusting the MaximumParallelInvocationsPerClient configuration. Upon method completion, the server utilizes the SendAsync method to transmit data back to the original caller.

However, we've encountered an issue where the data being sent back is quite large, currently it will send data **sequences** one after one,  resulting in a slowdown of our application. (For example, if I have 2 methods that take 1 minute, it will take me 2 minutes to complete).
In addition, if this process takes longer than 30 seconds (the default server timeout), the connection will reconnect.

Is my implementation correct? 
Are there any configurations or workarounds available that would enable concurrent data transmission to clients through a single connection?
Or we **must** use multiple connections to achieve it?

Thank you for your assistance.
Best regards.

**Comment by gfoidl** [comment]
@BrennanConroy please have a look here.

**Comment by quytrieu37** [comment]
## Diagnosis
- I believe the issue stems from each connection having a single "pipeline" for data transmission to the client. When this pipeline is occupied by one process, any additional processes are queued and sent to the client only after the previous ones have been completed.
- For the second issue, I think that because the connection is blocked by the current large message, subsequent keep-alive messages sent from the hub are queued and arrive later. As a result, the client doesn't receive the keep-alive message, triggering a reconnect on the client when exceeding the `DefaultServerTimeout`
